url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/parsertables/Nullable.java

```diff
@@ -0,0 +1,163 @@
+	there is a rule that contains the nonterminal on the left side and
+	no symbol on the right side (empty right side), expressing an optional
+	rule.
+	<p>
+	<b>Algorithm:</b><br>
+	Sort rules by size.
+	Search every rule that contain the nonterminal on left side.
+	Scan the left side of every rule until a terminal or a non-nullable
+	nonterminal appears. If found, rule is not nullable, else it is nullable.
+	If one of the rules that derive nonterminal is nullable, this nonterminal
+	is nullable.
+
+	@author (c) 2000, Fritz Ritzberger
+*/
+
+class Nullable extends Hashtable
+{
+	/** The special empty symbol. */
+	public static final String NULL = "";
+	
+	/**
+		Explores the nullability of all nonterminals in syntax. Provides nullability
+		as Boolean within this Map, key is nonterminal.
+		@param syntax the grammar to scan for nonterminals and their nullability.
+		@param nonterminals pre-built list of nonterminals.
+	*/
+	public Nullable(Syntax syntax, List nonterminals)
+		throws ParserBuildException
+	{
+		// loop nonterminals
+		Map done = new Hashtable(nonterminals.size());	// avoid endless loops
+		for (int i = 0; i < nonterminals.size(); i++)	{
+			String nt = (String) nonterminals.get(i);
+			checkNullability(syntax, nt, done);
+		}
+	}
+
+
+	/** Returns true if passed nonterminal is nullable, else false. */
+	public boolean isNullable(String nonterminal)	{
+		Boolean nullable = (Boolean) get(nonterminal);
+		return nullable.booleanValue();
+	}
+
+
+	private boolean checkNullability(Syntax syntax, String nonterm, Map done)
+		throws ParserBuildException
+	{
+		Boolean n = (Boolean) get(nonterm);
+		if (n != null)
+			return n.booleanValue();
+
+		if (done.get(nonterm) != null)
+			return false;	// endless recursion
+
+		done.put(nonterm, nonterm);	// avoid endless loops
+
+		// loop rules for an empty rule deriving this nonterminal
+		for (int j = 0; j < syntax.size(); j++)	{
+			Rule rule = syntax.getRule(j);
+			String nt = rule.getNonterminal();	// left side of derivation
+
+			if (nt.equals(nonterm) && rule.rightSize() <= 0)	// this rule derives the nonterminal and is empty
+				return putSymbol(nonterm, true);
+		}
+
+		// loop rules for nullable nonterminal sequences
+		for (int j = 0; j < syntax.size(); j++)	{
+			Rule rule = syntax.getRule(j);
+			String nt = rule.getNonterminal();	// left side of derivation
+
+			if (nt.equals(nonterm))	{	// this rule derives the nonterminal
+				boolean nullable = true;	// assume it is nullable
+
+				// then all symbols on right side must be nullable
+				for (int i = 0; nullable && i < rule.rightSize(); i++)	{
+					String symbol = rule.getRightSymbol(i);
+
+					if (Token.isTerminal(symbol))	{	// a terminal ends symbol-loop
+						nullable = false;
+					}
+					else
+					if (symbol.equals(nonterm) == false)	{	// do not search self
+						try	{
+							nullable = checkNullability(syntax, symbol, done);	// is nullable when this symbol is nullable
+						}
+						catch (Exception ex)	{
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)